cmd/bootstrap/cmd/final_list.go

```diff
@@ -0,0 +1,196 @@
+package cmd
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+
+	model "github.com/onflow/flow-go/model/bootstrap"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+var (
+	flagStakingNodesDir string
+)
+
+// finallistCmd represents the final list command
+var finalListCmd = &cobra.Command{
+	Use:   "finallist",
+	Short: "",
+	Long:  "",
+	Run:   finalList,
+}
+
+func init() {
+	rootCmd.AddCommand(finalListCmd)
+	addFinalListFlags()
+}
+
+func addFinalListFlags() {
+	// partner node info flag
+	finalListCmd.Flags().StringVar(&flagPartnerNodeInfoDir, "partner-infos", "", "path to a directory containing all parnter nodes details")
+	_ = finalListCmd.MarkFlagRequired("partner-infos")
+
+	// internal/flow node info flag
+	finalListCmd.Flags().StringVar(&flagInternalNodePrivInfoDir, "flow-infos", "", "path to a directory containing all internal/flow nodes details")
+	_ = finalListCmd.MarkFlagRequired("flow-infos")
+
+	// staking nodes dir containing staking nodes json
+	finalListCmd.Flags().StringVar(&flagStakingNodesDir, "staking-nodes", "", "path to a directory containing a JSON file of all staking nodes")
+	_ = finalListCmd.MarkFlagRequired("staking-nodes")
+}
+
+func finalList(cmd *cobra.Command, args []string) {
+	// read public partner node infos
+	log.Info().Msgf("reading parnter public node information: %s", flagPartnerNodeInfoDir)
+	partnerNodes := assemblePartnerNodesWithoutStake()
+
+	// read internal private node infos
+	log.Info().Msgf("reading internal/flow private node information: %s", flagInternalNodePrivInfoDir)
+	flowNodes := assembleInternalNodesWithoutStake()
+
+	log.Info().Msg("checking constraints on consensus/cluster nodes")
+	checkConstraints(partnerNodes, flowNodes)
+
+	log.Info().Msgf("reading staking contract node information: %s", flagStakingNodesDir)
+	stakingNodes := readStakingContractDetails()
+
+	// merge internal and partner node infos
+	mixedNodeInfos := mergeNodeInfos(flowNodes, partnerNodes)
+
+	// reconcile nodes from staking contract nodes
+	reconcileNodes(mixedNodeInfos, stakingNodes)
+
+	// write node-config.json with the new list of nodes to be used for the `finalize` command
+	writeJSON(fmt.Sprintf(flagOutdir, "node-config.json"), model.ToPublicNodeInfoList(mixedNodeInfos))
+}
+
+func reconcileNodes(nodes []model.NodeInfo, stakingNodes []model.NodeInfo) {
+	// check node count
+	if len(nodes) != len(stakingNodes) {
+		log.Error().Int("nodes", len(nodes)).Int("staked nodes", len(stakingNodes)).
+			Msg("staked node count does not match flow and parnter node count")
+	}
+
+	var nodesByID map[flow.Identifier]model.NodeInfo
+	for _, node := range nodes {
+		nodesByID[node.NodeID] = node
+	}
```

Comment:
@danuio let's make the check a bit more strict.

1. we need to go through each list, and check if there is missing node id on _each side_. If there is, print all the missing node id and then exist the program.

When we found missing node id during sporking, we could fix the data, and run this cmd again, until it passes.

2. We need to exist if the role doesn't match and print error

3. We could print warning if address or networking/staking key doesn't match, but still the program continue, as we mentioned, we will only strictly match node id for now.